### ISSUE TITLE
Explore poliastro-as-a-service options

### ISSUE THREAD

**Issue by Juanlu001** [poliastro/poliastro#444]
From [this comment by @ZachDischner](https://github.com/poliastro/poliastro/issues/156#issuecomment-418438770): 

> It'd be a nice separation to have a lightweight static web interface with poliastro and a js plotter so that all Poliastro runs is the backend calculations. Communicates with a service call so you could use a local service, or one in a cloud or on a server somewhere. Even serve the site statically on AWS...

And also related to the latest comments in https://github.com/poliastro/poliastro/issues/168.

According to [this list](https://www.g2crowd.com/categories/cloud-platform-as-a-service-paas?utf8=%E2%9C%93&order=popular), the most popular platform among the simplest ones I would try to support is AWS Lambda (FaaS). @ZachDischner already tried with no luck, see [this comment](https://github.com/poliastro/poliastro/issues/168#issuecomment-415499044):

> My unzipped env size ended up being around 375 MB.

Apart from this, other people might want to make poliastro work in:

* Google App Engine (PaaS) (which now [supports Python 3](https://cloud.google.com/blog/products/gcp/introducing-app-engine-second-generation-runtimes-and-python-3-7))
* Heroku/Dokku (PaaS) (good opportunity to create a Dockerfile)

We can't strive to cover all the possible use cases, but we should make sure that poliastro and all its dependencies work there. Giving some attention to https://github.com/poliastro/poliastro/issues/168 would help a lot in general, reduce deploy times, etc.

**Comment by robert-cronin** [comment]
I'm new to Poliastro so I might be off the mark here,

> It'd be a nice separation to have a lightweight static web interface with poliastro and a js plotter

Are you thinking something like a website that you can use to interact with Poliastro and display a plot of the entered orbits?
I also noticed your reference to the JPL Mission Design Tool, I guess this is similar to what you're after? 
https://github.com/poliastro/poliastro/issues/156#issuecomment-418431452

> We can't strive to cover all the possible use cases 

I think the 2D plots would be fun and easy to implement, so maybe that's where I can start.
If there's nobody else working on this, I might give it a shot!

**Comment by Juanlu001** [comment]
Thanks for chiming in @robert-cronin! I think there are three hallenges here:

1. Installing poliastro in some inexpensive platform. Of course we could rent a VPS and conda install everything there, but there are many advantages in using Function-as-a-Service or Platform-as-a-Service environments. It's not crucial, but it's important for scaling without paying a lot of money.
2. Developing some middle layer that exposes a REST API to a subset of poliastro functionality. This could already be used by a client like cURL, requests.py or similar. One has to figure out how to send plots like this.
3. Creating a web frontend for this API. It should be easy to maintain for future developers and, in the ideal case, be developed with modern JavaScript technologies that I hace no idea about :)

What do you think?

**Comment by Juanlu001** [comment]
I didn't think of it in depth, so up to you :) I would recommend working on
a Minimum Viable Product (MVP) first, something that requires minimal
effort but that already does "something".

**Comment by robert-cronin** [comment]
FYI I have started a dedicated repo, you can follow my progress here: [Poliastro App](https://github.com/robert-cronin/Poliastro-App)
I figured it might be cleaner if we keep it separate for now but we can merge it at later when it makes sense :)

**Comment by Juanlu001** [comment]
Perfect! Ping me if you need help :)

**Comment by sakethramanujam** [comment]
Has this issue been already taken care of?
If not I'd like to chip in and start building an MVP
1. In my current role, I've been using a lot of services on GCP and found it reliable to host APIs. 
2. GCP now has a product(currently in beta) called Cloud Run which is Lambda like service and also the app engine instances on GCP scale from 0 i.e., there's no cost involved when no one is using the service.
3. Talking of restAPI is there any spec that has already been mapped out?

**Comment by Juanlu001** [comment]
@sakethramanujam from what I understand, there were some limitations to run poliastro on AWS Lambda because of the size of the dependencies. If GCP Cloud Run is similar and has no such limitation, it would be great to see a prototype.

We did not agree on what functionality should be provided, let alone an API. We are open to suggestions :)

**Comment by Juanlu001** [comment]
https://oraas.orekit.space/

We had the idea first :)

**Comment by iamabhishek0** [comment]
@Juanlu001 @jorgepiloto  Are we looking for something like https://oraas.orekit.space/ ?
where users can select options from the dropdown?
I was thinking apart from having dropdown menus like the one in https://oraas.orekit.space/ , we can have a jupyter notebook or a Python shell where users can give the input and the output without having to setup Poliastro on the local machine.

**Comment by Phantsure** [comment]
that might be a good idea for those people who know poliastro but, not for those who just want to integrate poliastro in backend and want to design a good front end. Still your idea is nice.

**Comment by iamabhishek0** [comment]
Thanks @Phantsure, API's will also be created for the major functionalities of Poliastro.

**Comment by sakethramanujam** [comment]
@iamabhishek0 you mean an rpc kind of a service where the user would call the remote REST/whatever api and get back responses?

**Comment by iamabhishek0** [comment]
Yes, I was thinking of making REST API's using Flask/Django and hosting it all on Google App Engine/ AWS Lambda.

**Comment by sakethramanujam** [comment]
@iamabhishek0 sorry for the delay in reply. so you're proposing a flask wrapper around the existing api with the api payload being the parameters that are particular to a method. ?

**Comment by iamabhishek0** [comment]
@sakethramanujam Thanks for replying, Yes I was thinking of implementing the same, apart from a shell/jupyter-notebook, will that be a good idea?

**Comment by sakethramanujam** [comment]
@Juanlu001 your opinion please?

**Comment by xlxs4** [comment]
@astrojuanlu (ignore if you've seen this :D )

**Comment by astrojuanlu** [comment]
@xlxs4 A few thoughts on this:

Every year we get a lot of GSOC candidates that want to work on this project. And many of them send us proofs-of-concept that work! I've collected a few of them over the years.

However, there are some challenges to make poliastro-as-a-service a real thing:

- When I opened this issue https://github.com/poliastro/poliastro/issues/444#issue-356911663 there were problems to install poliastro in things like AWS Lambda. Perhaps this has changed in the past three years?
- It would be cool to try out to set up prototypes in AWS, GCP, Azure, Heroku, and whatever is trendy these days.
- **Importing poliastro is slow** because there is a lot of Numba JIT compilation behind the curtain, and we should address that.

Therefore, on one hand this is not a priority anymore, and on the other hand whoever wants to tackle this should present a compelling story about how to address these issues.

This is perhaps more of a _documentation_ project rather than a _coding_ project.

**Comment by ZachDischner** [comment]
Sorry I haven’t been very involved in this, I still think it is a cool
idea.

There indeed have been improvements to Lambda at the very least. The
original issue was that code size was too large (mostly due to
SciPy/Astropy). That may not be the case anymore. You can publish a docker
image for your lambda function, up to 10G in size. That might be the
cleanest tack to take here. Poliastro containers, runnable anywhere (I
still prefer Lambda ;-) ). If I get a chance I can try it out.


On March 24, 2021 at 10:58:02 AM, Juan Luis Cano Rodríguez (
***@***.***) wrote:

@xlxs4 <https://github.com/xlxs4> A few thoughts on this:

Every year we get a lot of GSOC candidates that want to work on this
project. And many of them send us proofs-of-concept that work! I've
collected a few of them over the years.

However, there are some challenges to make poliastro-as-a-service a real
thing:

   - When I opened this issue #444 (comment)
   <https://github.com/poliastro/poliastro/issues/444#issue-356911663>
   there were problems to install poliastro in things like AWS Lambda. Perhaps
   this has changed in the past three years?
   - It would be cool to try out to set up prototypes in AWS, GCP, Azure,
   Heroku, and whatever is trendy these days.
   - *Importing poliastro is slow* because there is a lot of Numba JIT
   compilation behind the curtain, and we should address that.

Therefore, on one hand this is not a priority anymore, and on the other
hand whoever wants to tackle this should present a compelling story about
how to address these issues.

This is perhaps more of a *documentation* project rather than a *coding*
project.

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub
<https://github.com/poliastro/poliastro/issues/444#issuecomment-805994780>,
or unsubscribe
